imports/ui/components/courses/events/course-events.js

```diff
@@ -119,3 +122,120 @@ Template.courseEventAdd.onRendered(function() {
 	eventCaption.on('mouseover mouseout', function(e) { toggleCaptionClass(e); });
 	instance.$('.event-caption-add-text').on('mouseover mouseout', function(e) { toggleCaptionClass(e); });
 });
+
+Template.deleteCourseEvents.helpers({
+	upcomingEvents() {
+		return Events.findFilter(
+			{ course: this.course._id
+			, after: minuteTime.get()
+			}
+		);
+	}
+});
+
+Template.deleteCourseEvents.events({
+	'click .js-show-events-delete-modal'(event, instance) {
+		instance.$('#deleteEventsModal').modal('show');
+	}
+});
+
+Template.deleteEventsModal.onCreated(function() {
+	this.busy(false);
+
+	this.state = new ReactiveDict();
+	this.state.setDefault(
+		{ selectedEvents: []
+		, allEventsSelected: false
+		, showDeleteConfirm: false
+		}
+	);
+
+	this.autorun(() => {
+		const cursor = Template.currentData().upcomingEvents;
+		const allEventsSelected = cursor.count() === this.state.get('selectedEvents').length;
+		this.state.set({ allEventsSelected });
+	});
+});
+
+Template.deleteEventsModal.helpers({
+	selectedEvents() {
+		return Template.instance().state.get('selectedEvents');
+	},
+
+	isSelected() {
+		return Template.instance().state.get('selectedEvents').find(e => e._id === this._id);
+	},
+
+	getState(state) {
+		return Template.instance().state.get(state);
+	},
+
+	numSelectedEvents() {
+		return Template.instance().state.get('selectedEvents').length;
+	},
+
+	disabledIfNoEventsSelected() {
+		if (Template.instance().state.get('selectedEvents').length === 0) {
+			return 'disabled';
+		}
+	}
+});
+
+Template.deleteEventsModal.events({
+	'click .js-toggle-all'(event, instance) {
+		let selectedEvents;
+		if (instance.state.get('allEventsSelected')) {
+			selectedEvents = [];
+		} else {
+			selectedEvents = Template.currentData().upcomingEvents.fetch();
+		}
+
+		instance.state.set({ selectedEvents });
+	},
+
+	'change input[type="checkbox"]'(event, instance) {
+		let selectedEvents = instance.state.get('selectedEvents');
+		if (event.target.checked) {
+			selectedEvents.push(this);
+		} else {
+			selectedEvents = selectedEvents.filter(e => e._id !== this._id);
+		}
+
+		instance.state.set({ selectedEvents });
+	},
+
+	'click .js-show-delete-confirm'(event, instance) {
+		instance.state.set('showDeleteConfirm', true);
+	},
+
+	'click .js-delete-events'(e, instance) {
+		instance.busy('deleting');
+		instance.state.get('selectedEvents').forEach((event, index, list) => {
+			Meteor.call('event.remove', event._id, (err) => {
+				if (index === list.length - 1) {
```

Comment:
This means error-handling is done for the last error only. All removals except the last could fail, and still the user would get a message that "n events were successfully deleted."

In my view it would be better to stop removing events at the first error returned and report it to the user. Keep a counter of events removed so the "deleted successfully" message can be show independently of the error message.

Pseudocode:
```
let countRemoved = 0;
each(event) {
    remove event
    if error {
      message "Failed removing" event "because" error
      break
    } else {
       removed++
    }   
}
if removed {
    message "Removed" countRemoved "events"
}
if not error {
    close overlay
}
```

This way the user gets feedback at the first error. The only problem I see with this is that we won't try to delete as many events as possible if one fails. In my view it is likely that removal of other events is likely to fail too so it's ok if we don't try. Errors should be uncommon here so it's ok to keep error-handling simple and just fail fast.

Reply:
I like the idea but how is this supposed to work with asynchronous method calls?

Reply:
Oops, indeed the calls are async. To get the sync behaviour we'd need to omit the result-handler. So it would read:

```
removed = 0
error = false
try {
  each event {
      remove event
      removed++
  }
} catch error {
     message "Failed removing" event "because" error
     error = true
}
if removed {
    message "Removed" countRemoved "events"
}
if not error {
    close overlay
}
```

Reply:
Doing it in-sync is much slower. But for this rare operation I don't think people care all that much. If we want async behaviour:

```
removed = 0
responses = 0
handler(error) {
  responses++
  if (error) {
    message "Failed removing" event "because" error
  } else {
    removed++
  }
  if responses == count events {
    if removed {
      message "Removed" countRemoved "events"
    }
    if removed == responses {
      close overlay
    }
  }
}

each event {
      remove event -> handler
}
```

Can't really say which version I prefer.